# Project

> **Learning goal:** By the end of this Capstone, you should be familiar with some of the ways to visually explore the data stored in __`DataFrame`__s.

Often when probing a new data set, it is invaluable to get high-level information about what the dataset holds. Earlier in this section we discussed using methods such as __`DataFrame.info()`__, __`DataFrame.head()`__, and __`DataFrame.tail()`__ to examine some aspects of a __`DataFrame`__. While these methods are critical, they are on their own often insufficient to get enough information to know how to approach a new dataset. This is where exploratory statistics and visualizations for datasets come in.

To see what we mean in terms of gaining exploratory insight (both visually and numerically), let's dig into one of the the datasets that come with the __`scikit-learn`__ library, the Boston Housing Dataset (though you will load it from a CSV file):

In [4]:
import pandas as pd

df = pd.read_csv('Data/housing_dataset.csv')
df.head()

# used to have Race as data, was removed
# has measurments of housing
# zillow - how mich is your house worth

# for each house in this dataset, we've captured information
# ZN - that's pretty huge, so it's like mansions
# INDUS - industry, factories
# in machine learning, a dummy variable is a way to convert a categorical value (usually string) into a number (like 'Male'/'Female' could be 1/0)
# Charles River - if it's next to the river or not
# NOX - pollution
# RM how big is the house
# AGE - the bigger, the older it is
# PTE - the smaller, the better
# LSTAT - economic status
# MEDV - the actual value of the home in millions of dollars


# so we have a lot of housing data and a lot of features

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO  LSTAT  MEDV  
0     15.3   4.98  24.0  
1     17.8   9.14  21.6  
2     17.8   4.03  34.7  
3     18.7   2.94  33.4  
4     18.7   5.33  36.2

This dataset contains information collected from the U.S Census Bureau concerning housing in the area of Boston, Massachusetts and was first published in 1978. The dataset has 14 columns:
 - **CRIM**:     Per-capita crime rate by town
 - **ZN**:       Proportion of residential land zoned for lots over 25,000 square feet
 - **INDUS**:    Proportion of non-retail business acres per town
 - **CHAS**:     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
 - **NOX**:      Nitric oxides concentration (parts per 10 million)
 - **RM**:       Average number of rooms per dwelling
 - **AGE**:      Proportion of owner-occupied units built prior to 1940
 - **DIS**:      Weighted distances to five Boston employment centres
 - **RAD**:      Index of accessibility to radial highways
 - **TAX**:      Full-value property-tax rate per \$10,000
 - **PTRATIO**:  Pupil-teacher ratio by town
 - **LSTAT**:    Percent of lower-status portion of the population
 - **MEDV**:     Median value of owner-occupied homes in \$1,000s
 
 One of the first methods we can use to better understand this dataset is `DataFrame.shape`:

In [2]:
df.shape # let's look at the shape of the data frame

(506, 13)

The dataset has 506 rows and 13 columns.

To get a better idea of the contents of each column we can use `DataFrame.describe`, which returns the maximum value, minimums value, mean, and standard deviation of numeric values in each columns, in addition to the quartiles for each column:

In [5]:
df.describe() # descriptive statistics of data, to get a goog picture of the data

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.593761   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.596783   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.647423   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO       LSTAT  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534   12.653063   
std     28.148861    2.105710    8.707259  168.537116    2.164946    7.141062   
min      2.900000    1.129600    1.000000  187.000000   12.600000    1.730000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000    6.950000   
50%     77.500000    3.207450    5.000000  330.000000   19.050000   11.360000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000   16.955000   
max    100.000000   12.126500   24.000000  711.000000   22.000000   37.970000   

             MEDV  
count  506.000000  
mean    22.532806  
std      9.197104  
min      5.000000  
25%     17.025000  
50%     21.200000  
75%     25.000000  
max     50.000000

Because dataset can have so many columns in them, it can often be useful to transpose the results of `DataFrame.describe` to better use them.
Note that you can also examine specific descriptive statistics for columns without having to invoke `DataFrame.describe`:

In [8]:
df['MEDV'].mean()

22.532806324110698

In [9]:
df['MEDV'].max()

50.0

In [10]:
df['AGE'].median()

77.5

In [11]:
# Now find the maximum value in df['AGE'].


Other information that you will often want to see is the relationship between different columns. You do this with the `DataFrame.groupby` method. For example, you could examine the average MEDV (median value of owner-occupied homes) for each value of AGE (proportion of owner-occupied units built prior to 1940):

In [35]:
df.groupby(['AGE'])['MEDV'].mean() # Panda lets us group the data by a specific criteria
# the MEDV value of the home, grouped by age
# we see that when it's an older home, it's cheaper. but in the middle we also see fluctuations, so we can't draw explicit conclusions.
# we can try to make an assumption and confirm/disconfirm it

AGE
2.9      26.600000
6.0      24.100000
6.2      23.400000
6.5      24.700000
6.6      24.750000
6.8      29.600000
7.8      23.250000
8.4      42.800000
8.9      24.800000
9.8      23.700000
9.9      31.100000
10.0     29.000000
13.0     24.400000
13.9     32.900000
14.7     24.600000
15.3     34.900000
15.7     42.300000
15.8     34.900000
16.3     25.200000
17.0     46.700000
17.2     22.600000
17.5     23.950000
17.7     33.100000
17.8     23.500000
18.4     25.150000
18.5     21.150000
18.8     29.100000
19.1     20.900000
19.5     20.600000
20.1     22.500000
           ...    
95.8     18.800000
96.0     13.500000
96.1     25.450000
96.2     27.966667
96.4     14.900000
96.6     11.700000
96.7     17.050000
96.8     50.000000
96.9     13.500000
97.0     14.033333
97.1     19.800000
97.2     17.100000
97.3     17.533333
97.4     24.133333
97.5     50.000000
97.7     19.600000
97.8     11.800000
97.9     18.250000
98.0      8.100000
98.1     10.400000
98.2     24.150000
98.3    

In [36]:
# Now try to find the median value for AGE for each value of MEDV.
df.groupby(['AGE'])['MEDV'].median()

AGE
2.9      26.60
6.0      24.10
6.2      23.40
6.5      24.70
6.6      24.75
6.8      29.60
7.8      23.25
8.4      42.80
8.9      24.80
9.8      23.70
9.9      31.10
10.0     29.00
13.0     24.40
13.9     32.90
14.7     24.60
15.3     34.90
15.7     42.30
15.8     34.90
16.3     25.20
17.0     46.70
17.2     22.60
17.5     23.95
17.7     33.10
17.8     23.50
18.4     25.15
18.5     21.15
18.8     29.10
19.1     20.90
19.5     20.60
20.1     22.50
         ...  
95.8     18.80
96.0     14.35
96.1     25.45
96.2     20.80
96.4     14.90
96.6     11.70
96.7     17.05
96.8     50.00
96.9     13.50
97.0     15.10
97.1     19.80
97.2     17.10
97.3     19.10
97.4     17.80
97.5     50.00
97.7     19.60
97.8     11.80
97.9     17.50
98.0      8.10
98.1     10.40
98.2     16.65
98.3     10.00
98.4     16.35
98.5     19.40
98.7     14.10
98.8     13.45
98.9     11.50
99.1     10.90
99.3     17.80
100.0    14.50
Name: MEDV, Length: 356, dtype: float64

You can also apply a lambda function to each element of a `DataFrame` column by using the `apply` method. For example, say you wanted to create a new column that flagged a row if more than 50 percent of owner-occupied homes were build before 1940:

In [37]:
df['AGE_50'] = df['AGE'].apply(lambda x: x > 50)

Once applied, you also see how many values returned true and how many false by using the `value_counts` method:

In [38]:
df['AGE_50'].value_counts()

True     359
False    147
Name: AGE_50, dtype: int64

You can also examine figures from the groupby statement you created earlier:

In [39]:
df.groupby(['AGE_50'])['MEDV'].mean()

AGE_50
False    26.693197
True     20.829248
Name: MEDV, dtype: float64

You can also group by more than one variable, such AGE_50 (the one you just created), CHAS (whether a town is on the Charles River), and RAD (an index measuring access to the Boston-area radial highways), and then evaluate each group for the average median home price in that group:

In [40]:
groupby_twovar = df.groupby(['AGE_50','RAD','CHAS'])['MEDV'].mean()

You can then see what values are in this stacked group of variables:

In [41]:
groupby_twovar

AGE_50  RAD   CHAS
False   1.0   0.0     24.666667
              1.0     50.000000
        2.0   0.0     33.300000
        3.0   0.0     26.505556
        4.0   0.0     25.376744
              1.0     32.900000
        5.0   0.0     26.302857
              1.0     46.000000
        6.0   0.0     23.575000
        7.0   0.0     28.563636
        8.0   0.0     29.220000
        24.0  0.0     20.766667
True    1.0   0.0     20.185714
        2.0   0.0     24.170588
        3.0   0.0     29.350000
              1.0     27.950000
        4.0   0.0     17.879661
              1.0     21.560000
        5.0   0.0     25.124638
              1.0     25.610000
        6.0   0.0     19.822222
        7.0   0.0     24.433333
        8.0   0.0     32.321429
              1.0     26.000000
        24.0  0.0     15.306612
              1.0     31.362500
Name: MEDV, dtype: float64

Let's take a moment to analyze these results in a little depth. The first row reports that communities with less than half of houses built before 1940, with a highway-access index of 1, and that are not situated on the Charles River have a mean house price of \$24,667 (1970s dollars); the next row shows that for communities similar to the first row except for being located on the Charles River have a mean house price of \$50,000.

One insight that pops out from continuing down this is that, all else being equal, being located next to the Charles River can significantly increase the value of newer housing stock. The story is more ambiguous for communities dominated by older houses: proximity to the Charles significantly increases home prices in one community (and that one presumably farther away from the city); for all others, being situated on the river either provided a modest increase in value or actually decreased mean home prices.

While groupings like this can be a great way to begin to interrogate your data, you might not care for the 'tall' format it comes in. In that case, you can unstack the data into a "wide" format:

In [42]:
groupby_twovar.unstack()

CHAS               0.0      1.0
AGE_50 RAD                     
False  1.0   24.666667  50.0000
       2.0   33.300000      NaN
       3.0   26.505556      NaN
       4.0   25.376744  32.9000
       5.0   26.302857  46.0000
       6.0   23.575000      NaN
       7.0   28.563636      NaN
       8.0   29.220000      NaN
       24.0  20.766667      NaN
True   1.0   20.185714      NaN
       2.0   24.170588      NaN
       3.0   29.350000  27.9500
       4.0   17.879661  21.5600
       5.0   25.124638  25.6100
       6.0   19.822222      NaN
       7.0   24.433333      NaN
       8.0   32.321429  26.0000
       24.0  15.306612  31.3625

In [43]:
# How could you use groupby to get a sense of the proportion 
# of residential land zoned for lots over 25,000 sq.ft., 
# the proportion of non-retail business acres per town, 
# and the distance of towns from employment centers in Boston?


It is also often valuable to know how many unique values a column has in it with the `nunique` method:

In [44]:
df['CHAS'].nunique()

2

Complementary to that, you will also likely want to know what those unique values are, which is where the `unique` method helps:

In [45]:
df['CHAS'].unique()

array([0., 1.])

You can use the `value_counts` method to see how many of each unique value there are in a column:

In [ ]:
df['CHAS'].value_counts()

Or you can easily plot a bar graph to visually see the breakdown:

In [ ]:
%matplotlib inline
df['CHAS'].value_counts().plot(kind='bar')

Note that the Jupyter magic command `%matplotlib inline` enables you to view the chart inline.

Let's pull back to the dataset as a whole for a moment. Two major things that you will look for in almost any dataset are trends and relationships. A typical relationship between variables to explore is the Pearson correlation, or the extent to which two variables are linearly related. The `corr` method will show this in table format for all of the columns in a `DataFrame`:

In [31]:
df.corr(method='pearson')
# is there anything in my dataset that is highly correlated with something else in my dataset?
# corr method
# example - how does pollution correlate with the value of the home - it's all negative
# so when the pollution goes up, the price goes down
# same with crime - more crime rate, cheaper homes

# we might want to look at correlations btw attributes, like pollution and factories
# multi colliniarity - when we find that variables are highly correlated


# so we want to look at our data before we even start doing things with it

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
CRIM     1.000000 -0.199458  0.404471 -0.055295  0.417521 -0.219940  0.350784   
ZN      -0.199458  1.000000 -0.533828 -0.042697 -0.516604  0.311991 -0.569537   
INDUS    0.404471 -0.533828  1.000000  0.062938  0.763651 -0.391676  0.644779   
CHAS    -0.055295 -0.042697  0.062938  1.000000  0.091203  0.091251  0.086518   
NOX      0.417521 -0.516604  0.763651  0.091203  1.000000 -0.302188  0.731470   
RM      -0.219940  0.311991 -0.391676  0.091251 -0.302188  1.000000 -0.240265   
AGE      0.350784 -0.569537  0.644779  0.086518  0.731470 -0.240265  1.000000   
DIS     -0.377904  0.664408 -0.708027 -0.099176 -0.769230  0.205246 -0.747881   
RAD      0.622029 -0.311948  0.595129 -0.007368  0.611441 -0.209847  0.456022   
TAX      0.579564 -0.314563  0.720760 -0.035587  0.668023 -0.292048  0.506456   
PTRATIO  0.288250 -0.391679  0.383248 -0.121515  0.188933 -0.355501  0.261515   
LSTAT    0.452220 -0.412995  0.603800 -0.053929  0.590879 -0.613808  0.602339   
MEDV    -0.385832  0.360445 -0.483725  0.175260 -0.427321  0.695360 -0.376955   
AGE_50   0.254574 -0.590769  0.516001  0.088659  0.597644 -0.164465  0.870348   

              DIS       RAD       TAX   PTRATIO     LSTAT      MEDV    AGE_50  
CRIM    -0.377904  0.622029  0.579564  0.288250  0.452220 -0.385832  0.254574  
ZN       0.664408 -0.311948 -0.314563 -0.391679 -0.412995  0.360445 -0.590769  
INDUS   -0.708027  0.595129  0.720760  0.383248  0.603800 -0.483725  0.516001  
CHAS    -0.099176 -0.007368 -0.035587 -0.121515 -0.053929  0.175260  0.088659  
NOX     -0.769230  0.611441  0.668023  0.188933  0.590879 -0.427321  0.597644  
RM       0.205246 -0.209847 -0.292048 -0.355501 -0.613808  0.695360 -0.164465  
AGE     -0.747881  0.456022  0.506456  0.261515  0.602339 -0.376955  0.870348  
DIS      1.000000 -0.494588 -0.534432 -0.232471 -0.496996  0.249929 -0.673813  
RAD     -0.494588  1.000000  0.910228  0.464741  0.488676 -0.381626  0.361191  
TAX     -0.534432  0.910228  1.000000  0.460853  0.543993 -0.468536  0.381395  
PTRATIO -0.232471  0.464741  0.460853  1.000000  0.374044 -0.507787  0.236216  
LSTAT   -0.496996  0.488676  0.543993  0.374044  1.000000 -0.737663  0.468146  
MEDV     0.249929 -0.381626 -0.468536 -0.507787 -0.737663  1.000000 -0.289750  
AGE_50  -0.673813  0.361191  0.381395  0.236216  0.468146 -0.289750  1.000000

Suppose you just wanted to look at the correlations between all of the columns and just one variable? Let's examine just the correlation between all other variables and the median value of the homes (MEDV). We will do this by accessing the column by index number:

In [32]:
corr = df.corr(method='pearson') # compute the corelation of every other variable
corr_with_homevalue = corr.iloc[-2]
corr_with_homevalue[corr_with_homevalue.argsort()[::-1]]
# we'll sort them from highest to lowest, which of my attributes are corelated/contribute the most to the price of the home?
# size of the house - it will be pricier, no matter in which neighborhood

MEDV       1.000000
RM         0.695360
ZN         0.360445
DIS        0.249929
CHAS       0.175260
AGE_50    -0.289750
AGE       -0.376955
RAD       -0.381626
CRIM      -0.385832
NOX       -0.427321
TAX       -0.468536
INDUS     -0.483725
PTRATIO   -0.507787
LSTAT     -0.737663
Name: MEDV, dtype: float64

In [34]:
corr = df.corr(method='pearson')
corr_with_homevalue = corr.iloc[-1]
corr_with_homevalue[corr_with_homevalue.argsort()[::-1]]

# let's look at the big negative correlations
# in areas where the socio-economic income is low, the price goes down
# on the other hand, living next to the river doesn't matter much

AGE_50     1.000000
AGE        0.870348
NOX        0.597644
INDUS      0.516001
LSTAT      0.468146
TAX        0.381395
RAD        0.361191
CRIM       0.254574
PTRATIO    0.236216
CHAS       0.088659
RM        -0.164465
MEDV      -0.289750
ZN        -0.590769
DIS       -0.673813
Name: AGE_50, dtype: float64

With the correlations arranged in descending order, it's easy to start to see some patterns. Correlating AGE with a variable we created from AGE is a trivial correlation. However, it is interesting to note that the percentage of older housing stock in communities strongly correlates with air pollution (NOX) and the proportion of non-retail business acres per town (INDUS); at least in 1978 metro Boston, older towns are more industrial.

Graphically, we can see the correlations using a heatmap from the Seaborn library:

In [ ]:
%matplotlib inline
import seaborn as sns
import numpy as np
sns.set(rc={'figure.figsize':(11, 8)})
# Generate a mask for the upper triangle / values above the identity diagonal
# Remove use of the mask below to see the "whole" heatmap
mask = np.zeros_like(corr, dtype=np.bool)
mask[np.triu_indices_from(mask)] = True
sns.heatmap(corr, cmap='coolwarm', annot=True, mask=mask);

Histograms are another valuable tool for investigating your data. For example, what is the overall distribution of prices of owner-occupied houses in the Boston area?

In [ ]:
import matplotlib.pyplot as plt
plt.hist(df['MEDV'])

The default bin size for the matplotlib histogram (essentially big of buckets of percentages that you include in each histogram bar in this case) is pretty large and might mask smaller details. To get a finer-grained view of the AGE column, you can manually increase the number of bins in the histogram:

In [ ]:
plt.hist(df['MEDV'], bins=50);

Seaborn has a somewhat more attractive version of the standard matplotlib histogram: the distribution plot. This is a combination histogram and kernel density estimate (KDE) plot (essentially a smoothed histogram):

In [ ]:
sns.distplot(df['MEDV'])

Another commonly used plot is the Seaborn jointplot, which combines histograms for two columns along with a scatterplot:

In [ ]:
sns.jointplot(df['RM'], df['MEDV'], kind='scatter') # we see that most homes have 6 rooms and are worth 20k

Unfortunately, many of the dots print over each other. You can help address this by adding some alpha blending, a figure that sets the transparency for the dots so that concentrations of them drawing over one another will be apparent:

In [ ]:
sns.jointplot(df['RM'], df['MEDV'], kind='scatter', alpha=0.3) 

Another way to see patterns in your data is with a two-dimensional KDE plot. Darker colors here represent a higher concentration of data points:

In [ ]:
sns.kdeplot(df['RM'], df['MEDV'], shade=True)

Note that while the KDE plot is very good at showing concentrations of data points, finer structures like linear relationships (such as the clear relationship between the number of rooms in homes and the house price) are lost in the KDE plot.

Finally, the pairplot in Seaborn allows you to see scatterplots and histograms for several columns in one table. Here we have played with some of the keywords to produce a more sophisticated and easier to read pairplot that incorporates both alpha blending and linear regression lines for the scatterplots.

In [ ]:
sns.pairplot(df[['RM', 'AGE', 'LSTAT', 'DIS', 'MEDV']], kind="reg", plot_kws={'line_kws':{'color':'red'}, 'scatter_kws': {'alpha': 0.1}})

Visualization is the start of the really cool, fun part of data science. So play around with these visualization tools and see what you can learn from the data!

> **Takeaway:** An old joke: “What does a data scientist seen when they look at a dataset? A bunch of numbers.” There is more than a little truth in that joke. Visualization is often the key to finding patterns and correlations in your data. While visualization cannot often deliver precise results, it can point you in the right direction to ask better questions and efficiently find value in the data.

In [ ]:
# you can't understand your data just by looking at numbers
# you should use plots to visualise your data